File: setup/backup-schedule.ts
import { spawn } from 'node:child_process';
import fs from 'node:fs/promises';
// after so many years of syncfs today I find
// that node:fs without /promise is non-promise-fs, not syncfs, it contains callback fs as async fs
import npfs from 'node:fs';
import path from 'node:path';
import dayjs from 'dayjs';
import dayjsUTCPlugin from 'dayjs/plugin/utc.js';
import yaml from 'yaml';
import * as oss from './alioss.ts';
import type { OSSConfig } from './alioss.ts';

dayjs.extend(dayjsUTCPlugin);
const configPath = path.resolve(process.env['FINE_CONFIG_DIR'] ?? '', 'backup.yml');
const config = yaml.parse(await fs.readFile(configPath, 'utf-8')) as {
    databases: string[],
    aliyunoss: OSSConfig & {
        path?: string,
    },
};

// add simply log to record details
// every backup operation generate a new log file, log file name is date,
// first line of log file is a success flag, avoid duplicate run in same day
// create a log() function capturing logs: string[]
function createlog(logs: string[]) {
    return (message: string) => {
        logs.push(`${dayjs().format('HH:mm:ss.SSS')} ${message}`);
    };
}

// return nothing: although there maybe error (how?), other backup operations should continue regardless of this
async function dumpDatabases(logs: string[]) {
    const log = createlog(logs);
    await Promise.all(config.databases.map(async item => {
        const [username, databaseName] = item.split(':').map(v => v.trim());
        console.log(`backup.ts: dump database ${item}`);
        log(`spawning pg_dump ${item}`);
        const child = spawn('pg_dump', ['--username', username, databaseName], { stdio: ['ignore', 'pipe', 'pipe'] });
        const content = await new Promise<string>(resolve => {
            let content = '';
            child.stdout.on('data', data => {
                content += data; // this seems to be string
            });
            child.stderr.on('data', data => {
                log(`pg_dump ${item} unexpected stderr: ${data}`);
            });
            child.on('error', error => {
                log(`pg_dump ${item} error? ${error}`);
                resolve(null); // ignore error and continue other backup work
            });
            child.on('close', async (code, signal) => {
                // document says one of code and signal will not be null
                // if you code??signal, it will become null for code=0, so use signal??code
                log(`pg_dump ${item} returned with ${signal ?? code}, received stdout ${content.length} bytes`);
                resolve(content);
            });
        });
        if (content) {
            log(`pg_dump ${item} write to /data/base/${databaseName}.sql`);
            try {
                // this by default overwrites file if exists, which is good
                await fs.writeFile(`/data/base/${databaseName}.sql`, content);
            } catch (error) {
                log(`pg_dump ${item} write ${databaseName}.sql error, skip, ${error}`);
            }
        } else {
            log(`pg_dump ${item} empty stdout?`);
        }
    }));
}

// return result file, return null for not ok
// // save the content and tar tf to check valid: await fs.writeFile('test.tar.xz', bundleFiles(logs));
async function bundle(logs: string[]): Promise<Buffer<ArrayBuffer>> {
    const log = createlog(logs);
    console.log(`backup.ts: tar`);
    log('spawning tar');
    // no specific data needed in stdout or stderr, use inherit
    const child = spawn('tar', ['cJf', '-', '/data'], { stdio: ['ignore', 'pipe', 'pipe'] });
    return await new Promise<Buffer<ArrayBuffer>>(resolve => {
        const chunks = [];
        child.stdout.on('data', data => {
            chunks.push(data);
        });
        child.stderr.on('data', data => {
            log(`tar stderr: ${data}`);
        });
        child.on('error', error => {
            log(`tar error? abort, ${error}`);
            console.log(`backup.ts: tar error? check logs for detail`);
            resolve(null);
        });
        child.on('close', async (code, signal) => {
            const data = Buffer.concat(chunks);
            log(`tar returned with ${signal ?? code}, received stdout ${data.length} bytes`);
            resolve(data);
        });
    });
}

// return true for ok, false for now ok
async function upload(time: dayjs.Dayjs, logs: string[], content: Buffer<ArrayBuffer>): Promise<boolean> {
    const filename = `fine-backup-${time.format('YYYYMMDD-HHmmss')}.tar.xz`;
    console.log(`backup.ts: upload ${filename}`);
    const result = await oss.upload(config.aliyunoss, { time, logs, filename, content });
    if (!result.ok) {
        console.log(`backup.ts: upload not ok, check logs for detail`);
    }
    return result.ok;
}

const logsDirectory = path.resolve(process.env['BACKUP_LOGS_DIR'] ?? '');
async function backupOnce(time: dayjs.Dayjs) {
    console.log(`backup.ts: start backup`);
    let ok = false;
    const logs: string[] = [];
    await dumpDatabases(logs);
    const bundledContent = await bundle(logs);
    if (bundledContent) {
        ok = await upload(time, logs, bundledContent);
    }
    logs.unshift(ok ? 'ok' : 'no');

    const backupLogFileName = path.resolve(logsDirectory, `backup-${time.format('YYYYMMDD-HHmmss')}.log`);
    console.log(`backup.ts: save log to ${backupLogFileName}`);
    const resultlogs = logs.map(r => r.trim() + '\n').join('');
    try {
        await fs.writeFile(backupLogFileName, resultlogs);
    } catch (error) {
        console.log(resultlogs);
        console.log(`backup.ts: write log file error?`, error);
        console.log(`backup.ts: abort because cannot maintain backup logs to correctly schedule`);
        process.exit(1);
    }
}
// await backupOnce(dayjs.utc());

async function schedule() {
    console.log('backup.ts: start scheduling');

    let sleeping = false;
    // use request shutdown not direct shutdown to avoid interupting network operation,
    // interupt pgdump or tar is ok actually, but no need to distinguish them
    let shutdownRequested = false;
    const requestShutdown = () => {
        if (sleeping) {
            // ...but when sleeping you need to exit directly, or else have to wait 1 hour
            console.log('backup.ts: shutdown requested, exit');
            process.exit(0);
        } else {
            console.log('backup.ts: requesting shutdown');
            shutdownRequested = true;
        }
    };
    process.on('SIGINT', requestShutdown);
    process.on('SIGTERM', requestShutdown);

    // let mocktime = dayjs.utc();
    while (true) {
        if (shutdownRequested) {
            console.log('backup.ts: shutdown requested, exit (2)');
            process.exit(0);
        }
        sleeping = true;
        // this is sleep, if you forget
        await new Promise<void>(resolve => setTimeout(() => resolve(), 3600_000)); // mock: 1_000)); ((
        sleeping = false;
        // wakeup and find it's time and find no related success log for this date, initiate backup

        let time = dayjs.utc(); // mock: mocktime = mocktime.add(1, 'hour'); to achieve add 1 hour per second
        if (time.hour() < 18) { continue; } // use hour >= 18 to effectly allow 6 times of retry if some error happens

        let filenames: string[];
        try {
            filenames = await fs.readdir(logsDirectory);
        } catch (error) {
            console.log('backup.ts: readdir error', error);
            console.log(`backup.ts: abort because cannot maintain backup logs to correctly schedule`);
            process.exit(1);
        }
        let ok = false;
        const todayFileNames = filenames
            // reverse sort filenames, the ok file normally is the last file if there are multiple files
            .filter(n => n.startsWith(`backup-${time.format('YYYYMMDD')}`)).sort((n1, n2) => n2.localeCompare(n1));
        for (const filename of todayFileNames) {
            try {
                const file = npfs.createReadStream(path.join(logsDirectory, filename), {
                    encoding: 'utf-8',
                    // NOTE the range is inclusive
                    // if file length is less than this, will return less content not throw error
                    start: 0, end: 1,
                });
                if ((await file.toArray()).join('') == 'ok') { ok = true; break; }
            } catch (error) {
                console.log(`backup.ts: read previous log file ${filename} error, skip`, error);
            }
        }
        if (ok) { continue; }

        // mock content for test, adjust  vvv  this number to control success rate
        // const mockok = Math.random() > 0.5 ? 'ok' : 'no';
        // console.log(`mock content backup-${time.format('YYYYMMDD-HHmmss')}.log: ${mockok}`);
        // await fs.writeFile(path.resolve(logsDirectory, `backup-${time.format('YYYYMMDD-HHmmss')}.log`), mockok);
        // continue;
        await backupOnce(time);
    }
}
await schedule();

File: setup/backup-schedule.py
#!/usr/local/bin/python -u
#                  NOTE ^^ this -u is for unbufferred output, which is required if you want to docker logs python entrypoint print output
import os, datetime, tarfile, subprocess, tomllib, pathlib, shutil, signal
from time import sleep

# 1. program: pack index.js, configs and certificates into fine-program-{datetime}.tar.xz
def backup_program(log, time):
    result_filename = f'/result/fine-program-{time.strftime('%Y%m%d-%H%M%S')}.tar.xz'
    with tarfile.open(result_filename, 'w:xz') as f:
        # recursive by default,
        # no need to strip prefix in tarfile, they will mee same file structure when restore
        f.add('/data/main')
    print(f'backup.py: create {result_filename}')
    for filename in os.listdir('/result'):
        if filename.startswith('fine-program-') and filename != result_filename[8:]:
            log(f'remove previous file {filename}')
            os.remove(f'/result/{filename}')

# 2. database: pg_dump and pack into fine-database-{datetime}.tar.xz
def backup_database(log, time):
    pathlib.Path('/data/base').mkdir(parents=True, exist_ok=True)
    with open('/data/main/configs/backup.toml', 'rb') as config_file:
        config = tomllib.load(config_file)
        for item in config['databases']:
            username, databasename = item.split(':')
            log(f'pgdump user {username}, database {databasename}')
            result_filename = f'/data/base/{databasename}.sql'
            child = subprocess.run(['pg_dump', '--username', username, '-f', result_filename, databasename], capture_output=True)
            if child.stdout: log(f'pgdump {databasename} stdout: {child.stdout}')
            if child.stderr: log(f'pgdump {databasename} stderr: {child.stderr}')
            print(f'backup.py: pg_dump {databasename} return code {child.returncode}')
    result_filename = f'/result/fine-database-{time.strftime('%Y%m%d-%H%M%S')}.tar.xz'
    with tarfile.open(result_filename, 'w:xz') as f:
        f.add('/data/base')
    print(f'backup.py: create {result_filename}')
    for filename in os.listdir('/result'):
        if filename.startswith('fine-database-') and filename != result_filename[8:]:
            log(f'remove previous file {filename}')
            os.remove(f'/result/{filename}')

# 3. logs: pack certbot, database, core, backup logs into fine-logs.tar.xz
def backup_logs(log, time):
    # normalize certbot log file name
    certbot_log_dir = pathlib.Path('/data/logs/certbot')
    certbot_log_dir.mkdir(parents=True, exist_ok=True)
    certbot_staging_log_dir = pathlib.Path('/data/certbot-logs-staging')
    for filename in os.listdir(certbot_staging_log_dir):
        filepath = certbot_staging_log_dir / filename
        with open(filepath) as file:
            logtime = file.read(19)
        if len(logtime) != 19:
            continue # there is empty file? skip empty file
        logtime = datetime.datetime.strptime(logtime, '%Y-%m-%d %H:%M:%S')
        # / has higher priority than +, I guess
        expect_filename = 'letsencrypt-' + logtime.strftime('%Y%m%d-%H%M%S') + '.log'
        if filename == expect_filename:
            continue
        expect_filepath = certbot_log_dir / expect_filename
        if not expect_filepath.exists():
            filepath.copy(expect_filepath, preserve_metadata=True)
            log(f'certbot log: normalize {filename} to {expect_filename}')
    result_filename = f'/result/fine-logs.tar.xz'
    # w:xz overwrite existing file
    with tarfile.open(result_filename, 'w:xz') as f:
        f.add('/data/logs')
    print(f'backup.py: create {result_filename}')

# 4. additional: pack public into fine-public.tar.xz, node_modules into fine-node_modules.tar.xz
def backup_additional(log, time):
    result_filename = f'/result/fine-public.tar.xz'
    with tarfile.open(result_filename, 'w:xz') as f:
        f.add('/data/public')
    print(f'backup.py: create {result_filename}')
    result_filename = f'/result/fine-node_modules.tar.xz'
    with tarfile.open(result_filename, 'w:xz') as f:
        f.add('/data/node_modules')
    print(f'backup.py: create {result_filename}')

def backup_once(time):
    ok = False
    print(f'backup.py: start backup at {time.strftime('%Y-%m-%d %H:%M:%S')}')
    with open(f'/data/logs/backup/backup-{time.strftime('%Y%m%d-%H%M%S')}.log', 'a') as logfile:
        log = lambda content: logfile.write(f"{datetime.datetime.now(tz=datetime.UTC).strftime('%Y%m%dT%H%M%SZ')} {content}\n")
        try:
            backup_program(log, time)
            backup_database(log, time)
            backup_logs(log, time)
            backup_additional(log, time)
            # and finally rsync, use a dedicated sync command is how
            # you keep network operations parallel when they are contained inside a cli
            print(f'backup.py: rsync')
            child = subprocess.run([
                '/work/async',
                '-c', '/data/main/configs/backup.toml',
                'rsync',
                '--local', '/result',
                '--compare-hash',
            ], capture_output=True, env=dict(os.environ, RUST_LOG='r#async=trace'))
            if child.stdout: log(f'rsync stdout: {child.stdout}')
            if child.stderr: log(f'rsync stderr: {child.stderr}')
            ok = child.returncode == 0
            child_output = child.stdout.decode().strip()
            print('\n'.join([f'  rsync: {r}' for r in child_output.split('\n')]))
            print(f'backup.py: rsync return code {child.returncode}')
        except Exception as error:
            log(f'panic! {error}')
            print(f'backup.py: panic! {error}')
    print(f'backup.py: complete backup at {datetime.datetime.now(tz=datetime.UTC).strftime('%Y-%m-%d %H:%M:%S')}')
    return ok
# backup_once(datetime.datetime.now(tz=datetime.UTC))

def schedule():
    # this is different from certbot-renew-schedule,
    # that run randomly twice a day, regardless of command result,
    # this I expect always run successfully once a day, and don't need randomize
    # so the strategy is, always sleep to next minute = 0, check logs/last-ok, if today not done, run
    
    sleeping = False
    shutdown_requested = False
    def request_shutdown(s, f):
        if sleeping:
            print('backup.py: interrupting sleep...')
            raise SystemExit(0)
        else:
            print('backup.py: requesting shutdown...')
            shutdown_requested = True
    signal.signal(signal.SIGINT, request_shutdown)
    signal.signal(signal.SIGTERM, request_shutdown)

    startup_time = datetime.datetime.now(datetime.UTC)
    print(f'backup.py: start at {startup_time.strftime('%Y-%m-%d %H:%M:%S')}')

    while True:
        if shutdown_requested:
            print('backup.py: shutdown requested, exit')
            exit()
        now = datetime.datetime.now(datetime.UTC)
        # sleep to next time point with minute = 0
        try:
            sleeping = True
            sleep((60 - now.minute) * 60)
        except SystemExit:
            print('backup.py: shutdown requested in sleep, exit')
            exit(0)
        finally:
            sleeping = False
        # use hour >= 18 to effectively allows 6 times of retry if some error happens
        if now.hour < 18: continue
        # check last ok time
        last_ok_path = '/data/logs/backup/last-ok'
        try:
            with open(last_ok_path) as f:
                last_ok_time = datetime.datetime.strptime(f.read(), '%Y-%m-%d %H:%M:%S').replace(tzinfo=datetime.UTC)
        except Exception as error:
            print(f'backup.py: failed to load last ok time from last-ok file?', error)
        # today ok
        if now.date() == last_ok_time.date(): continue

        ok = backup_once(now) # <-- main action is here
        with open(last_ok_path, 'w') as f:
            f.write(now.strftime('%Y-%m-%d %H:%M:%S'))
        # a good complete log is written in backup once
schedule()
